quisp/modules/Application.cc

```diff
@@ -60,185 +24,174 @@ Application::Application() { generatePacket = nullptr; }
-            EV << "Just one lonely connection setup request will be sent from " << myAddress << " to " << endnode_destination_address << "\n";
-            ConnectionSetupRequest *pk = new ConnectionSetupRequest();
-            pk->setActual_srcAddr(myAddress);
-            pk->setActual_destAddr(endnode_destination_address);
-            pk->setDestAddr(myAddress);
-            pk->setSrcAddr(myAddress);
-            pk->setNumber_of_required_Bellpairs(number_of_resources);
-            pk->setKind(7);
-            scheduleAt(simTime(), pk);
-          }
-          break;
-        case 2:  // let's all mambo!
-          // while ((endnode_destination_address = getOneRandomEndNodeAddress()) == myAddress);
-          endnode_destination_address = getOneRandomEndNodeAddress();
-          EV << "My connection setup request will be sent from " << myAddress << " to " << endnode_destination_address << "\n";
-          ConnectionSetupRequest *pk = new ConnectionSetupRequest();
-          pk->setActual_srcAddr(myAddress);
-          pk->setActual_destAddr(endnode_destination_address);
-          pk->setDestAddr(myAddress);
-          pk->setSrcAddr(myAddress);
-          pk->setNumber_of_required_Bellpairs(number_of_resources);  // required bell pairs
-          pk->setKind(7);
-          // to avoid conflict
-          scheduleAt(simTime() + exponential(0.00001 * myAddress), pk);
-          break;
-      }
+    return;
+  }
+
+  my_address = getParentModule()->par("address");
+  is_e2e_connection = par("EndToEndConnection");
+  number_of_resources = par("NumberOfResources");
+  num_measure = par("num_measure");
+
+  other_end_node_addresses = storeEndNodeAddresses();
+
+  if (!is_e2e_connection) {
+    return;
+  }
+
+  int traffic_pattern = par("TrafficPattern");
+
+  if (traffic_pattern == 0) {
+    EV_INFO << "EndToEndConnection is set true. but no traffic pattern specified; proceeding with no traffic\n";
+    return;
+  }
+
+  // just one connection
+  if (traffic_pattern == 1) {
+    int initiator_address = par("LoneInitiatorAddress");
+    if (my_address == initiator_address) {
+      int endnode_dest_addr = getOneRandomEndNodeAddress();
+      EV_INFO << "Just one lonely connection setup request will be sent from " << my_address << " to " << endnode_dest_addr << "\n";
+      ConnectionSetupRequest *pk = createConnectionSetupRequest(endnode_dest_addr, number_of_resources);
+      scheduleAt(simTime(), pk);
     }
+    return;
   }
+
+  // let's all mambo!
+  // Each EndNode makes exactly one connection.
+  // this means that some nodes will be receivers of more than one connection, at random.
+  if (traffic_pattern == 2) {
+    int endnode_dest_addr = getOneRandomEndNodeAddress();
+    EV_INFO << "My connection setup request will be sent from " << my_address << " to " << endnode_dest_addr << "\n";
+    ConnectionSetupRequest *pk = createConnectionSetupRequest(endnode_dest_addr, number_of_resources);
```

Comment:
I don't want to think about the detail of the packet creation.
so I move this into `createConnectionSetupRequest` method